src/lang.rs

```diff
@@ -383,1255 +791,1737 @@ fn is_keyword(c: char) -> bool {
-    peek!(ws!(pair!(low_filter, alt!(tag!("AND") | tag!("and"))))) >>
-    res: map!(
-        return_error!(SyntaxErrors::InvalidBooleanExpression.into(),
-            separated_pair!(ws!(low_filter),
-            ws!(tag!("AND")),
-            ws!(low_filter))), |(l,r)|Search::And(vec![l, r])
-    ) >> (res)
-));
-
-named!(filter_explicit_or<Span, Search>, do_parse!(
-    peek!(ws!(pair!(low_filter, alt!(tag!("OR") | tag!("or"))))) >>
-    res: map!(
-        return_error!(SyntaxErrors::InvalidBooleanExpression.into(),
-            separated_pair!(ws!(mid_filter),
-            ws!(tag!("OR")),
-            ws!(mid_filter))), |(l,r)|Search::Or(vec![l, r])
-    ) >> (res)
-));
-
-// Top level:
-// Look for OR
-// Look for AND
-// Implicit AND
-
-/* A list of things is implicitly ANDed together */
-named!(filter_implicit_and<Span, Search>, map!(
-    separated_nonempty_list!(multispace, low_filter),
-    // An empty keyword would match everything, so there's no reason to
-    |mut v| {
-        v.retain( | k | ! k.no_op());
-        Search::And(v)
-    }
-));
-
-named!(filter_atom<Span, Search>, alt_complete!(
-    map!(quoted_string, |s| Search::Keyword(Keyword::new_exact(s.to_string()))) |
-    map!(keyword, |s| Search::Keyword(Keyword::new_wildcard(s.trim_matches('*').to_string())))
-));
-
-named!(high_filter<Span, Search>, alt_complete!(
-  filter_explicit_or |
-  mid_filter
-));
-
-named!(mid_filter<Span, Search>, alt_complete!(
-  filter_explicit_and |
-  low_filter
-));
-
-named!(low_filter<Span, Search>, alt_complete!(
-  filter_not
-  | filter_atom
-  | delimited!(ws!(tag!("(")), high_filter, ws!(tag!(")")))
-));
-
-named!(filter_not<Span, Search>, map!(
-        ws!(preceded!(tag!("NOT"), low_filter)),
-        |k|Search::Not(Box::new(k))
-    )
-);
-
-named!(filter_expr<Span, Search>,
-    ws!(exact!(alt_complete!(
-        filter_explicit_or |
-        filter_explicit_and |
-        ws!(filter_implicit_and)
-))));
-
-named!(query_component<Span, Span>, alt_complete!(
-        recognize!(quoted_string) |
-        tag!("AND") |
-        tag!("NOT") |
-        tag!("OR") |
-        tag!("(") |
-        tag!(")") |
-        recognize!(keyword)
-    )
-);
+///
+fn comp_op(input: Span) -> IResult<Span, ComparisonOp> {
+    alt((
+        tag("==").map(|_| ComparisonOp::Eq),
+        tag("!=").map(|_| ComparisonOp::Neq),
+        tag(">=").map(|_| ComparisonOp::Gte),
+        tag("<=").map(|_| ComparisonOp::Lte),
+        tag(">").map(|_| ComparisonOp::Gt),
+        tag("<").map(|_| ComparisonOp::Lt),
+    ))(input)
+}
+
+fn unary_op(input: Span) -> IResult<Span, UnaryOp> {
+    tag("!").map(|_| UnaryOp::Not).parse(input)
+}
+
+fn muldiv_op(input: Span) -> IResult<Span, ArithmeticOp> {
+    alt((
+        tag("*").map(|_| ArithmeticOp::Multiply),
+        tag("/").map(|_| ArithmeticOp::Divide),
+    ))(input)
+}
 
-named!(end_of_query<Span, Span>, alt!(
-    ws!(tag!("|")) | ws!(eof!())
-));
+fn term(input: Span) -> IResult<Span, Expr> {
```

Comment:
can you add a short comment to this about what it parses? (and to any other parsing functions that don't have very obvious names)

Reply:
added